src/analyses/mutexAnalysis.ml

```diff
@@ -146,38 +119,57 @@ struct
       let create_protected protected = `Lifted2 protected
     end
 
-    let add ctx (l:Mutexes.elt*bool) =
-      let s,m = ctx.local in
-      let s' = Lockset.add l s in
-      match Addr.to_mval (fst l) with
-      | Some mval when MutexTypeAnalysis.must_be_recursive ctx mval ->
-        (s', Multiplicity.increment (fst l) m)
-      | _ -> (s', m)
-
-    let remove' ctx ~warn l =
-      let s, m = ctx.local in
-      let rm s = Lockset.remove (l, true) (Lockset.remove (l, false) s) in
-      if warn &&  (not (Lockset.mem (l,true) s || Lockset.mem (l,false) s)) then M.warn "unlocking mutex (%a) which may not be held" Addr.pretty l;
-      match Addr.to_mval l with
-      | Some mval when MutexTypeAnalysis.must_be_recursive ctx mval ->
-        let m',rmed = Multiplicity.decrement l m in
-        if rmed then
-          (rm s, m')
+    let add ctx ((addr, rw): AddrRW.t): D.t =
+      match addr with
+      | Addr mv ->
+        let (s, m) = ctx.local in
+        let s' = MustLocksetRW.add_mval_rw (mv, rw) s in
+        let m' =
+          if MutexTypeAnalysis.must_be_recursive ctx mv then
+            MustMultiplicity.increment mv m
+          else
+            m
+        in
+        (s', m')
+      | NullPtr ->
+        M.warn "locking NULL mutex";
+        ctx.local
+      | StrPtr _
+      | UnknownPtr -> ctx.local
+
+    let remove' ctx ~warn (addr: Addr.t): D.t =
+      match addr with
+      | StrPtr _
+      | UnknownPtr -> ctx.local
+      | NullPtr ->
+        if warn then
+          M.warn "unlocking NULL mutex";
+        ctx.local
+      | Addr mv ->
+        let (s, m) = ctx.local in
+        if warn && not (MustLocksetRW.mem_mval mv s) then
+          M.warn "unlocking mutex (%a) which may not be held" Mval.pretty mv;
+        if MutexTypeAnalysis.must_be_recursive ctx mv then (
+          let (m', rmed) = MustMultiplicity.decrement mv m in
+          if rmed then
+            (* TODO: don't repeat the same semantic_equal checks *)
+            (* TODO: rmed per lockset element, not aggregated *)
+            (MustLocksetRW.remove_mval mv s, m')
+          else
+            (s, m')
+        )
         else
-          (s, m')
-      | _ -> (rm s, m)
+          (MustLocksetRW.remove_mval mv s, m) (* TODO: should decrement something if may be recursive? *)
```

Comment:
No, not really, as nothing was incremented in the first place. Our `MutexTypeAnalysis` is flow-insensitive.
The only thing to consider is when that information changes during iteration. However, since the order is must here, no values where the multiplicity is > 0 for non-must-recursive mutexes should be present at fixpoint.